specification/draft-2/README.md

```diff
@@ -0,0 +1,687 @@
+TODO: Intro
+
+ - CWL specifies a way of describing runnable command line tools and workflows.
+ - CWL descriptions are serialized as either YAML or JSON.
+ - A CWL description contains a formal declaration of the inputs and outputs of described process (using Apache Avro schema).
+
+
+# Core specification
+
+CWL specification comes in two parts:
+
+ - Core spec allows description of basic tools and workflows
+ - Extensions add features that enable describing more complex tools and workflows
+
+If using an extension feature, it needs to be explicitly listed in the requirements section
+of the description document. Vendors might also add features incompatible with rest of CWL -
+in this case, they should be explicitly listed as requirements with full URI as value of `class` property.
+
+
+## Command line tools
+
+Basic tool example:
+
+```yaml
+class: CommandLineTool
+label: Grep
+description: Filter lines of text file by regular expression.
+inputs:
+  - id: pattern
+    type: string
+    binding:
+      position: 1
+  - id: in_file
+    type: File
+    streamable: true
+    binding:
+      position: 2
+outputs:
+  - id: out_file
+    type: File
+    streamable: true
+    binding:
+      glob: output.txt
+baseCmd: grep
+stdout: output.txt
+successCodes: [0, 1]
+```
+
+Above is defined a command line tool labeled "grep" with two required inputs
+(a file and a string) and one file output.
+
+Input objects contain a `binding` section which describes how the arguments
+are passed onto the command line. In this case, `pattern` and `in_file` are
+positional arguments and `pattern` is passed first, followed by `in_file`.
+Identifiers for inputs and outputs must be unique across the document.
+
+The `binding` section of the output object contains a single property `glob`
+with a glob pattern describing how to recognize file generated by the tool.
+
+Additionally, `baseCmd` and `stdout` properties describe the executable name
```

Comment:
@scottfrazer I'm not sure how stdin and stdout handling is any different?  Unless you are assuming that the script gets piped to the shell.  I'm proposing that the shell command is placed in the shell invocation command line like this:

```/bin/sh -c "first_cmd | second_cmd"```

then it connects the shell's stdin to `first_cmd` and  `second_cmd` to the shell's stdout.  The runtime would continue to handle stdin/stdout redirection around the `sh` process exactly the same way it does now.

@ntijanic I think the only reasonable/safe approach is to quote everything by default, and provide a per-input flag to disable quoting explicitly.  Example:

```yaml
shell: true
arguments:
  - first_cmd
  - "my file"
  - {value: "|", shellquote: false}
  - second_cmd
```

It builds the command line as a list of tokens (existing behavior), quotes each token (unless `shellquote: false`) then joins with spaces to get the literal string that will be passed to the shell.

This would result in an argv something like:

```json
["/bin/sh", "-c", "first_cmd my\\ file | second_cmd"]
```